backend/src/services/hsm/hsm-service.ts

```diff
@@ -0,0 +1,349 @@
+import grapheneLib from "graphene-pk11";
+
+import { getConfig } from "@app/lib/config/env";
+import { logger } from "@app/lib/logger";
+
+import { HsmModule, RequiredMechanisms } from "./hsm-types";
+
+type THsmServiceFactoryDep = {
+  hsmModule: HsmModule;
+};
+
+const USER_ALREADY_LOGGED_IN_ERROR = "CKR_USER_ALREADY_LOGGED_IN";
+const WRAPPED_KEY_LENGTH = 32 + 8; // AES-256 key + padding
+
+export type THsmServiceFactory = ReturnType<typeof hsmServiceFactory>;
+
+type SyncOrAsync<T> = T | Promise<T>;
+type SessionCallback<T> = (session: grapheneLib.Session) => SyncOrAsync<T>;
+
+export const withSession = async <T>(
+  { module, graphene }: HsmModule,
+  callbackWithSession: SessionCallback<T>
+): Promise<T> => {
+  const appCfg = getConfig();
+
+  let session: grapheneLib.Session | null = null;
+  try {
+    if (!module) {
+      throw new Error("PKCS#11 module is not initialized");
+    }
+
+    // Create new session
+    const slot = module.getSlots(appCfg.HSM_SLOT);
+    // eslint-disable-next-line no-bitwise
+    if (!(slot.flags & graphene.SlotFlag.TOKEN_PRESENT)) {
+      throw new Error("Slot is not initialized");
+    }
+
+    for (let i = 0; i < 10; i += 1) {
```

Comment:
Why is this for-loop and subsequent promise / sleep needed again? I can't imagine there's no cleaner way to do this?

Reply:
There can only be one session open per slot. Let's say that the user uses horizontal scaling for their instance, and 10 instances start up at the same time. They will all attempt to open a session at the same time, which will throw an error for most of those instances, because a session will already be open.

To fix this, we wait and try to get the session, I'm open for suggestions on how you'd improve this, since there's no native way to "listen" for open sessions.